Orange/widgets/utils/itemmodels.py

```diff
@@ -76,11 +85,29 @@ def columnCount(self, parent=QModelIndex()):
         return 0 if parent.isValid() else max(map(len, self._table), default=0)
 
     def data(self, index, role=Qt.DisplayRole):
-        if index.isValid() and role == Qt.DisplayRole:
-            try:
-                return self[index.row()][index.column()]
-            except IndexError:
-                pass
+        if not index.isValid():
+            return
+        try:
+            value = self[index.row()][index.column()]
+        except IndexError:
+            return
+        if role == Qt.DisplayRole:
+            if (isinstance(value, Number) and
+                    not (isnan(value) or isinf(value) or isinstance(value, Integral))):
+                absval = abs(value)
+                strlen = len(str(int(absval)))
+                value = '{:.{}{}}'.format(value,
+                                          2 if absval < .001 else
+                                          3 if strlen < 2 else
+                                          1 if strlen < 5 else
+                                          0 if strlen < 6 else
+                                          3,
+                                          'f' if strlen < 6 and absval > .001 else 'e')
+            return str(value)
```

Comment:
`value` may already be a string here. Perhaps `return` inside `if`?

On a related note, I'm concerned about the speed of this conversion if the table is huge.